### ISSUE TITLE
unbounded channel usage

### ISSUE THREAD

**Issue by flxo** [tqwewe/kameo#3]
Hello,

thanks for sharing the project. I'm tired of typing my own message passing with channels, an enum and tasks over and over again ;-)

What's the rational behind using `unbounded` channels? I debugged channel usage *a lot* and always came to the conclusion that a stuck task/actor that is backpressured from somewhere is easier to find than something like a memory leak in an unbounded channel. 

thanks

@flxo

**Comment by tqwewe** [comment]
Hi @flxo thanks for taking the time to check out Kameo!

**Tldr;** api & maintainance simplicity, since a battle tested system like erlang uses unbounded channels - but open to adding support for bounded channels!

---

In an older project I previously did use bounded channels, however for simplicity I decided on unbounded channels after researching if elixir has any bounds on their mailboxes, which apparently don't if I'm not mistaken.

I came to the conclusion that backpressure can be simulated by using `actor_ref.send(msg).await`, since this itself awaits for the message to be completed.

I could consider adding support for using bounded channnels instead of unbounded, but this means the API would probably be expanded to have in addition to `.send` `.send_async`: `.try_send`, `.try_send_async`, `.send_async_timeout` which might add confusion to end users having so many methods.

**Comment by flxo** [comment]
> Hi @flxo thanks for taking the time to check out Kameo!
> 
> **Tldr;** api & maintainance simplicity, since a battle tested system like erlang uses unbounded channels - but open to adding support for bounded channels!
> 
> In an older project I previously did use bounded channels, however for simplicity I decided on unbounded channels after researching if elixir has any bounds on their mailboxes, which apparently don't if I'm not mistaken.

I totally agree that looking at Elixir/Erlang and following their approach can't be wrong in this context. This also lowers the bar for developers who used that before.

> I came to the conclusion that backpressure can be simulated by using `actor_ref.send(msg).await`, since this itself awaits for the message to be completed.

Hm. This `await` makes a subtle difference. Haven't checked the docs but I assume that if the oneshot receiver is dropped the receiving actor would just ignore the result when they try to send the result. 

Awaiting the message to be fully processed gives a lot of control, but might throttle, because the actor tasks can't make progress in parallel. I think that in the scenario where there's a continuous stream of messages from one actor to another, the classic backpressure with some buffering can help regarding throughput. 

> I could consider adding support for using bounded channnels instead of unbounded, but this means the API would probably be expanded to have in addition to `.send` `.send_async`: `.try_send`, `.try_send_async`, `.send_async_timeout` which might add confusion to end users having so many methods.

So the `try_` variants would return something like `Err(MailboxFull)`? There could also be the behaviour that the "oldest" message in the queue is discarded if the channel is full. Now it get's confusing. ;-).

Exposing the channel fill states would also be a plus. When working with bounded channels there always needs to be a decision about the size. This is in reality an educated guess and often wrong. I often saw people heedlessly increasing channels sizes to hide bugs (mostly deadlocks). Adjusting channel capacities must be done after checking real world conditions.

**Comment by flxo** [comment]
> Yeah very likely, I'm not sure what the best solution might be, but would be worth considering support for bounded channels. ~Sadly you can't even see the current size of an unbounded channel to print warnings if the size starts growing too high too~ just saw there is a method 🤦

Don't worry. Its very very [new](https://github.com/tokio-rs/tokio/blob/master/tokio/CHANGELOG.md).

**Comment by tqwewe** [comment]
I'll close this for now since unbounded channels seems to work well for the BEAM runtime. But I'm definitely fine to reopen this if its a feature needed by users

**Comment by tqwewe** [comment]
Reopening, since I think bounded channels would be more ideal.
